### PR TITLE
Added multilabel evaluation 

### PR DESCRIPTION
**Pull request by mihail911**
Added support for multilabel evaluation in classification_model base class. @BoltzmannBrain @scottpurdy @subutai

### CODE REVIEW

File: fluent/experiments/baseline_experiment.py [hunk 1]
```diff
@@ -45,13 +45,14 @@
 import numpy
 import os
 import time
+import collections
```
Comment (BoltzmannBrain) [inline fluent/experiments/baseline_experiment.py]: Imports in alpha order please, as noted in our [coding style guide](https://github.com/numenta/nupic/wiki/Python-Style-Guide#imports).

File: fluent/experiments/baseline_experiment.py [hunk 2]
```diff
@@ -159,15 +166,37 @@ def run(args):
   print "Reading in data and preprocessing."
   preprocessTime = time.time()
   texter = TextPreprocess()
-  samples, labels = readCSV(dataPath, 2, [3])  # [3] is single class, range(3,6) is multi-class
+
+  if args.useMultiClass:
+    rawSamples, labels = readCSV(dataPath, 2, range(3,6))  # [3] is single class, range(3,6) is multi-class
```
Comment (BoltzmannBrain) [inline fluent/experiments/baseline_experiment.py]: comment no longer needed

File: tests/unit/classification_model_test.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+    topLabels = model._getTopLabels(inferenceResult, numLabels=1)
+    self.assertTrue(numpy.allclose(topLabels, numpy.array([2])))
+
+    topLabels = model._getTopLabels(inferenceResult, numLabels=2)
+    self.assertTrue(numpy.allclose(topLabels, numpy.array([2,0])))
+
+    topLabels = model._getTopLabels(inferenceResult, numLabels=4)
+    self.assertTrue(numpy.allclose(topLabels, numpy.array([2,0,1,3])))
+
```
Comment (BoltzmannBrain) [inline tests/unit/classification_model_test.py]: Please use two blank lines between functions.

File: fluent/experiments/baseline_experiment.py [hunk 4]
```diff
@@ -45,13 +45,14 @@
...
 import time
+import collections
 
 from fluent.utils.csv_helper import readCSV
 from fluent.utils.data_split import KFolds
 from fluent.utils.text_preprocess import TextPreprocess
 
 
-def runExperiment(model, patterns, labels, idxSplits):
+def runExperiment(model, patterns, labels, idxSplits, sampleLabelList):
```
Comment (BoltzmannBrain) [inline fluent/experiments/baseline_experiment.py]: add `sampleLabelList` to docstring params

File: tests/unit/classification_model_test.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import unittest
+import numpy
+import pandas
+
+from fluent.models import classify_random_sdr, classify_endpoint, \
```
Comment (BoltzmannBrain) [inline tests/unit/classification_model_test.py]: Import the classes explicitly so you don't have to repeatedly call e.g. `classify_random_sdr.ClassificationModelRandomSDR()` throughout.

File: tests/unit/classification_model_test.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    self.assertEquals(labels[1], output[1])
+
+  def testClassifyRandomSDRHard(self):
+    """Tests simple classification with multiple labels for `randomSDR`."""
+    samples =[['cat'], ['fat']]
+    labels = numpy.array([0, 1])
+    model = classify_random_sdr.ClassificationModelRandomSDR()
+
+    patterns = []
+    [patterns.append(model.encodePattern(samples[idx])) for idx in range(2)]
```
Comment (BoltzmannBrain) [inline tests/unit/classification_model_test.py]: Delete line 118 and just use the list comprehension... `patterns = [model.encodePattern(samples[idx]) for idx in xrange(2)]`. And doesn't matter here b/c range is so small, but `xrange()` is best.

File: tests/unit/classification_model_test.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+    for idx, p in enumerate(patterns):
+      model.trainModel(p, labels[idx])
+
+    output = [model.testModel(p, 2) for p in patterns]
+
+    self.assertTrue(numpy.allclose(output[0], numpy.array([0,1])))
```
Comment (BoltzmannBrain) [inline tests/unit/classification_model_test.py]: Please add failure messages to all assertions.

File: fluent/models/classification_model.py [hunk 8]
```diff
@@ -98,6 +98,19 @@ def _densifyPattern(self, bitmap):
...
+    """Returns the `n` most frequent labels in k nearest neighbors.
+
+    @param inferenceResult (numpy.array) Stores frequency of each category
+
+    @param numLabels (int) Return this number of most frequent labels
+          within top k
+
+    @return (numpy.array)  Return list of top `numLabels` labels
+
+  """
```
Comment (BoltzmannBrain) [inline fluent/models/classification_model.py]: indent

File: fluent/models/classification_model.py [hunk 9]
```diff
@@ -98,6 +98,19 @@ def _densifyPattern(self, bitmap):
       densePattern[i] = 1.0
     return densePattern
 
+  def _getTopLabels(self, inferenceResult, numLabels=1):
+    """Returns the `n` most frequent labels in k nearest neighbors.
+
+    @param inferenceResult (numpy.array) Stores frequency of each category
```
Comment (BoltzmannBrain) [inline fluent/models/classification_model.py]: Please follow the same docstring formatting as in the other methods.

File: fluent/models/classification_model.py [hunk 10]
```diff
@@ -98,6 +98,19 @@ def _densifyPattern(self, bitmap):
       densePattern[i] = 1.0
     return densePattern
```
Comment (BoltzmannBrain) [inline fluent/models/classification_model.py]: Two blank lines between methods.

File: fluent/models/classify_fingerprint.py [hunk 11]
```diff
@@ -107,7 +107,6 @@ def testModel(self, sample):
     Note: to return multiple winner classifications, modify the return statement
     accordingly.
     """
-    (tokenLabel, _, _, _) = self.classifier.infer(
+    (tokenLabel, inferenceResult, _, _) = self.classifier.infer(
```
Comment (BoltzmannBrain) [inline fluent/models/classify_fingerprint.py]: no longer using `tokenLabel`

File: fluent/experiments/baseline_experiment.py [hunk 12]
```diff
@@ -175,17 +204,23 @@ def run(args):
 
   print "Encoding the data."
   encodeTime = time.time()
-  patterns = [model.encodePattern(s) for s in samples]
+  patterns = []
+  for s in samples:
+    patterns.append((model.encodePattern(s[0]), s[1]))
```
Comment (BoltzmannBrain) [inline fluent/experiments/baseline_experiment.py]: Why do away w/ list comprehension? Use `patterns = [model.encodePattern(s[0], s[1]) for s in samples]`

File: fluent/experiments/baseline_experiment.py [hunk 13]
```diff
@@ -159,15 +166,37 @@ def run(args):
...
   labelReference = list(set(labels))
   labels = numpy.array([labelReference.index(l) for l in labels], dtype="int8")
-  samples = [texter.tokenize(sample,
+
+  # Responsible for generating a list of list of labels for the
+  # gold standard labels
+  sampleLabelMapping = collections.defaultdict(list)
+  for idx, s in enumerate(rawSamples):
+    sampleLabelMapping[s].append(labels[idx])
+  sampleLabelList = [sampleLabelMapping[s] for s in rawSamples]
```
Comment (BoltzmannBrain) [inline fluent/experiments/baseline_experiment.py]: It's better to move this logic out to its own function, also changing the if-statement bodies in lines 170-173 above. The other experiment runner is better organized this way, i.e. [`Runner.setupData()`](https://github.com/numenta/nupic.fluent/blob/master/fluent/experiments/runner.py#L76).

File: fluent/models/classification_model.py [hunk 14]
```diff
@@ -98,6 +98,19 @@ def _densifyPattern(self, bitmap):
...
+          within top k
+
+    @return (numpy.array)  Return list of top `numLabels` labels
+
+  """
+    return inferenceResult.argsort()[::-1][:numLabels]
+
 
   @staticmethod
   def winningLabels(labels, n=3):
```
Comment (subutai) [inline fluent/models/classification_model.py]: Does the docstring for winningLabels need to be updated?

Comment (BoltzmannBrain) [inline fluent/models/classification_model.py]: yes

File: fluent/models/classification_model.py [hunk 15]
```diff
@@ -108,7 +121,7 @@ def winningLabels(labels, n=3):
     """
     labelCount = Counter(labels).most_common()
     maxCount = 0
-    for c in labelCount:  ## TODO: better way to do this?
+    for c in labelCount:  ## TODO: better way to do this? -- most_common() should return a reverse sorted list so this is unnecessary
```
Comment (BoltzmannBrain) [inline fluent/models/classification_model.py]: nice catch... so if it's unnecessary then remove it :wink:

File: fluent/experiments/baseline_experiment.py (hunk 13 above)
Comment (BoltzmannBrain) [inline fluent/experiments/baseline_experiment.py]: I'm also finding a bug here running the real data. When the text sample is just "n", `sampleLabelList` has a list of many 8's, e.g. `patterns[956]` for q50 returns a tuple that claims the sample has 15 classifications, which is [8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8].

File: fluent/experiments/baseline_experiment.py [hunk 16]
```diff
@@ -159,15 +166,37 @@ def run(args):
...
+    samples = []
+    for sample in rawSamples:
+      samples.append((texter.tokenize(sample,
                              ignoreCommon=100,
                              removeStrings=["[identifier deleted]"],
-                             correctSpell=True)
-             for sample in samples]
-  # samples = [texter.tokenize(sample) for sample in samples]
+                             correctSpell=True),
+                             len(sampleLabelMapping[sample])))
```
Comment (BoltzmannBrain) [inline fluent/experiments/baseline_experiment.py]: Keeping a count along with the data is a bit hacky. I'd prefer the second item in the `samples` tuple (and subsequently the `patterns` tuple) be the list of the actual label indices, e.g.

    >>> print samples[0][1]
    [0, 7]

Alternatively keep these label lists in a separate data structure -- i.e. list of lists.

File: fluent/experiments/baseline_experiment.py [hunk 17]
```diff
@@ -175,17 +204,23 @@ def run(args):
...
   if args.train:
-    training(model, [(p, labels[i]) for i, p in enumerate(patterns)])
+    training(model, [(p[0], labels[i]) for i, p in enumerate(patterns)])
 
   if args.test:
-    results = testing(model, [(p, labels[i]) for i, p in enumerate(patterns)])
+    evalSet = []
+    for idx, p in enumerate(patterns):
+      goldLabels = sampleLabelList[idx]
+      evalSet.append((p[0], goldLabels, p[1]))
```
Comment (BoltzmannBrain) [inline fluent/experiments/baseline_experiment.py]: The revisions per my comment above regarding the labels lists should make this cleaner; i.e. get rid of `goldLabels` variable.

File: fluent/experiments/baseline_experiment.py [hunk 18]
```diff
@@ -202,7 +237,8 @@ def run(args):
     for k in xrange(args.kFolds):
       print "Training and testing for CV fold {0}.".format(k)
       kTime = time.time()
-      trialResults = runExperiment(model, patterns, labels, partitions[k])
+      trialResults = runExperiment(model, patterns, labels, partitions[k],
+                                   sampleLabelList)
```
Comment (BoltzmannBrain) [inline fluent/experiments/baseline_experiment.py]: Again, if `patterns` just has the labels lists in each item's tuple, `sampleLabelList` can be eliminated.

File: fluent/models/classification_model.py [hunk 19]
```diff
@@ -132,6 +145,7 @@ def evaluateResults(self, classifications, references, idx): ## TODO: evaluation
                                               ints or None, and items in actual
                                               classifications list are ints.
     @param references       (list)            Classification label strings.
+    @param idx              (list)            Indices from unknown `partitions` variable
```
Comment (BoltzmannBrain) [inline fluent/models/classification_model.py]: wrap lines at 80 characters

File: fluent/models/classification_model.py [hunk 20]
```diff
@@ -186,10 +200,14 @@ def calculateAccuracy(classifications):   ## TODO: just get numpy arrays passed
       raise ValueError("Classification lists must have same length.")
 
     actual = numpy.array(classifications[1])
-    predicted = numpy.array([c[0] for c in classifications[0]])  ## TODO: multiclass; this forces evaluation metrics to consider only the first predicted classification
+    predicted = numpy.array([c for c in classifications[0]])  ## TODO: multiclass; this forces evaluation metrics to consider only the first predicted classification
```
Comment (BoltzmannBrain) [inline fluent/models/classification_model.py]: Is list comprehension here still necessary? We're not grabbing the first element anymore...
Regardless, delete the TODO note.

File: fluent/models/classification_model.py [hunk 21]
```diff
@@ -186,10 +200,14 @@ def calculateAccuracy(classifications):   ## TODO: just get numpy arrays passed
...
 
-    return (actual == predicted).sum() / float(len(actual))
+    accuracy = 0.0
+    for idx in range(len(actual)):
```
Comment (BoltzmannBrain) [inline fluent/models/classification_model.py]: use `xrange()`

File: fluent/models/classification_model.py [hunk 22]
```diff
@@ -198,16 +216,19 @@ def calculateConfusionMatrix(classifications, references):  ## TODO: just get nu
       raise ValueError("Classification lists must have same length.")
 
     actual = numpy.array(classifications[1])
-    predicted = numpy.array([c[0] for c in classifications[0]])  ## TODO: multiclass; this forces evaluation metrics to consider only the first predicted classification
+    predicted = numpy.array([c for c in classifications[0]])  ## TODO: multiclass; this forces evaluation metrics to consider only the first predicted classification
```
Comment (BoltzmannBrain) [inline fluent/models/classification_model.py]: see comment on line 203

File: fluent/models/classification_model.py [hunk 23]
```diff
@@ -198,16 +216,19 @@ def calculateConfusionMatrix(classifications, references):  ## TODO: just get nu
...
 
     total = len(references)
     cm = numpy.zeros((total, total+1))
     for i, p in enumerate(predicted):
       if p is not None:
-        cm[actual[i]][p] += 1
+        for aLabel in actual[i]:
+          for pLabel in p:
+            cm[aLabel][pLabel] += 1
```
Comment (BoltzmannBrain) [inline fluent/models/classification_model.py]: Not calculating the CM correctly here; this requires the order of actual labels to match the order of the predicted labels for correct classifications. E.g. actual = [0,4,7] and predicted = [7, 0, 4] would yield three incorrect classifications with the above logic.

File: fluent/models/classify_endpoint.py [hunk 24]
```diff
@@ -137,7 +137,7 @@ def testModel(self, sample, numberCats=1, metric="overlappingAll"):
     for cat, catBitmap in self.categoryBitmaps.iteritems():
       distances[cat] = self.client.compare(sampleBitmap, catBitmap)
 
-    return self.winningLabels(distances, numberCats=1, metric="overlappingAll")  ## TODO: how do we handle return values of []? or len(winners)<numberCats?
+    return self.winningLabels(distances, numberCats=numLabels, metric="overlappingAll")  ## TODO: how do we handle return values of []? or len(winners)<numberCats?
```
Comment (BoltzmannBrain) [inline fluent/models/classify_endpoint.py]: Let's get consistent with the variable names... please change `numberCats` to `numLabels`.

File: fluent/models/classify_fingerprint.py [hunk 25]
```diff
@@ -107,7 +107,6 @@ def testModel(self, sample):
...
       self._densifyPattern(sample["bitmap"]))
-    ## TODO: get list of closest classifications, not just the winner; implement this in base class
-    return [tokenLabel]
+    return self._getTopLabels(inferenceResult, numLabels) #Return `n` most-frequent classifications
```
Comment (BoltzmannBrain) [inline fluent/models/classify_fingerprint.py]: Don't put comments at the end of a line. I may do this occasionally for TODOs, but that's just a personal reminder to come back and change the code asap.
And this comment is unnecessary anyway. Return info should be reflected in the docstring...

File: tests/__init__.py [hunk 26]
```diff
@@ -0,0 +1 @@
+__author__ = 'numenta'
```
Comment (BoltzmannBrain) [inline tests/__init__.py]: Why is this here? Just keep this file empty.

File: tests/unit/classification_model_test.py [hunk 27]
```diff
@@ -0,0 +1,183 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import unittest
```
Comment (BoltzmannBrain) [inline tests/unit/classification_model_test.py]: alpha order please

**Comment by BoltzmannBrain** [comment]
@mihail911 done with this round of review, left specific comments throughout. As you make the revisions, please be sure to revise the unit tests accordingly. Some overall comments:
- What about experiment.py?? This is the experiment we want running asap.
- Use list comprehensions over for loops with append b/c they're much faster; you don't need to load the append attribute and call it as a function each time.
- For unit/integration tests, please include failure messages with the assertions.
- After a fix that satisfies a TODO comment, remove the comment.

File: fluent/models/classification_model.py (hunk 21 above)
Comment (scottpurdy) [inline fluent/models/classification_model.py]: better:

```
for actualValue, predictedValue in zip(actual, predicted):
```

File: fluent/models/classification_model.py (hunk 22 above)
Comment (scottpurdy) [inline fluent/models/classification_model.py]: also, avoid trailing comments like this. Usually better to put the comment on the line before

File: fluent/models/classification_model.py [hunk 28]
```diff
@@ -198,16 +216,19 @@ def calculateConfusionMatrix(classifications, references):  ## TODO: just get nu
       raise ValueError("Classification lists must have same length.")
 
     actual = numpy.array(classifications[1])
-    predicted = numpy.array([c[0] for c in classifications[0]])  ## TODO: multiclass; this forces evaluation metrics to consider only the first predicted classification
+    predicted = numpy.array([c for c in classifications[0]])  ## TODO: multiclass; this forces evaluation metrics to consider only the first predicted classification
 
     total = len(references)
     cm = numpy.zeros((total, total+1))
     for i, p in enumerate(predicted):
```
Comment (scottpurdy) [inline fluent/models/classification_model.py]: can probably replace this with `for a, p in zip(actual, predicted)` as shown above

File: tests/unit/classification_model_test.py [hunk 29]
```diff
@@ -0,0 +1,183 @@
...
+import numpy
+import pandas
+
+from fluent.models import classify_random_sdr, classify_endpoint, \
+                    classify_fingerprint, classification_model
+
+# Unit tests to add:
```
Comment (scottpurdy) [inline tests/unit/classification_model_test.py]: If there are comments like this then put them in a doc string. In this case it would be the module doc string (between copyright and first import) but in other cases it could be a class or function doc string.

File: tests/unit/classification_model_test.py [hunk 30]
```diff
@@ -0,0 +1,183 @@
...
+  def testClassifyEndpoint(self):
+    """Tests sample classification with single label for `endpoint`."""
+    samples =[['cat'], ['fat']]
+    labels = numpy.array([0, 1])
+    model = classify_endpoint.ClassificationModelEndpoint()
+
+    patterns = []
+    [patterns.append(model.encodePattern(samples[idx])) for idx in range(2)]
+
+    for idx, p in enumerate(patterns):
```
Comment (scottpurdy) [inline tests/unit/classification_model_test.py]: Looks like there are a few cases in the tests where you could use zip rather than enumerate to simplify loops

**Comment by mihail911** [comment]
@BoltzmannBrain Have updated code as per everyone's suggestions

File: fluent/models/classification_model.py [hunk 31]
```diff
@@ -126,12 +140,15 @@ def evaluateResults(self, classifications, references, idx): ## TODO: evaluation
...
-                                              ints or None, and items in actual
-                                              classifications list are ints.
+    @param classifications  (list)            Two lists: (0) predictions and
+                                              (1) actual classifications. Items
+                                              in the predictions list are lists
+                                              of ints or None, and items in
+                                              actual classifications list are
+                                              ints.
     @param references       (list)            Classification label strings.
+    @param idx              (list)            Indices from unknown `partitions`
```
Comment (BoltzmannBrain) [inline fluent/models/classification_model.py]: don't put open questions in docstrings, they're supposed to be helpful... so if you can't figure it out just ask :smile:

**Comment by BoltzmannBrain** [comment]
Included in #91.
